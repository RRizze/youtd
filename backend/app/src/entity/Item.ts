import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  OneToOne,
  JoinColumn,
  ManyToMany,
  JoinTable
} from 'typeorm';
import { Rarity } from './Rarity';
import { Ability } from './Ability';

@Entity()
export class Item {
  @PrimaryGeneratedColumn()
  id: number;
  
  @Column({ type: 'varchar', length: 200 })
  itemName: string;

  @Column()
  goldcost: number;

  @Column({ type: 'varchar', length: 70 })
  icon: string;

  @Column()
  waveLvlDrop: number;

  @Column({ nullable: true })
  rarityId: number;

  @OneToOne(type => Rarity)
  @JoinColumn()
  rarity: Rarity;

  @Column({ type: 'text' })
  itemDescription: string;

  @Column({ type: 'text' })
  specials: string;

  @ManyToMany(type => Ability)
  @JoinTable()
  abilities: Ability[]

}