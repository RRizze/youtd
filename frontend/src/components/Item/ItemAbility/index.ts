export { default } from "./item-ability";
