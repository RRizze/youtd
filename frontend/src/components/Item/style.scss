@import '../../styles/mixins';
.item {
  min-width: 320px;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  width: 100%;
  position: relative;
  padding: 40px 2px 10px 2px;
  box-shadow: 1px 5px 8px #1c1821;

  .item__icon {
    width: 64px;
    height: 64px;
  }
}

.item__table-header {
  color: #ffffff;
  background-color: #1c1821;
  padding: 2px 3px;
  font-size: 16px;
}

.item__name {
  padding-bottom: 10px;
  padding-top: 10px;
  font-size: 18px;
  letter-spacing: 1px;
}

.item__general-info {
  font-size: 14px;
  width: 50%;
  padding-right: 3%;

  @include mw(375) {
    width: 100%;
  }
}

.item__table {
  width: 100%;

  tr:nth-child(odd) {
    background: #382f40;
  }
}

.item__main-info {
  width: 50%;

  @include mw(375) {
    width: 100%;
    padding-top: 20px;
  }
}

.item__specials, .item__abilities {
  padding-top: 10px;
}

.item__main-info-title {
  font-size: 18px;
  letter-spacing: 1px;
  color: #ffffff;
  display: inline-block;
  padding-bottom: 5px;
}

.main-info-title {
  color: #ffffff;
  background-color: #1c1821;
  padding: 2px 3px;
  font-size: 15px;
  display: inline-block;
}

