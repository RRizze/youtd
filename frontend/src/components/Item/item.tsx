import React from "react";
import "./style.scss";
import ItemAbility from "./ItemAbility";
import { IItemNORM } from "../../models/item";

const Item = ({
  item_name,
  icon,
  item_id,
  item_description,
  goldcost,
  wave_lvl_drop,
  rarity_name,
  specials,
  abilities,
}: IItemNORM) => {
  let ability;
  let special;

  if (typeof abilities === "object") {
    ability = <ul className="abilities-list">
      {abilities.map((a) => (
        <li key={a.id} className="abilities-list__item ability">
          <ItemAbility
            id={a.id}
            ability_name={a.ability_name}
            ability_description={a.ability_description}
            lvl_bonus={a.lvl_bonus}
            ability_active={a.ability_active} />
        </li>
      ))}
    </ul>;
  } else {
    ability = <p>{abilities}</p>;
  }

  if (typeof specials === "object") {
    special = <ul className="specials-list">
      {specials.map((s, idx) => (
        <li key={s + idx}>{s}</li>
      ))}
    </ul>;
  } else {
    special = <p>{specials}</p>;
  }

  return (
    <div className={"item"} data-id={item_id}>
      <div className="item__general-info">
        <div className="item__header">
          <figure>
            <img className="item__icon" src={icon} alt="item icon" />
            <figcaption><h2 className="item__name">{item_name}</h2></figcaption>
          </figure>
        </div>
        <div className="item__features">
          <table className="item__table">
            <thead>
              <tr>
                <th colSpan={2} className="item__table-header">General Info</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th>Rarity</th>
                <td className={`${rarity_name}`}>{rarity_name}</td>
              </tr>
              <tr>
                <th>Goldcost</th>
                <td>{goldcost}</td>
              </tr>
              <tr>
                <th>Wave Lvl Drop</th>
                <td>{wave_lvl_drop}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div className="item__main-info">
        <h2 className="item__main-info-title">Main Info</h2>
        <div className="item__description">
          <h3 className="main-info-title">Item Description</h3>
          <p>{item_description}.</p>
        </div>
        <div className="item__specials">
          <h3 className="main-info-title">Specials: </h3>
          {special}
        </div>
        <div className="item__abilities">
          <h3 className="main-info-title">Abilities: </h3>
          {ability}
        </div>
      </div>
    </div>
  );
};

export default Item;
