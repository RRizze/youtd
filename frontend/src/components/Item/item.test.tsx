import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Item from "./item";

describe("Item tests", () => {
  it("render correctly", () => {
    const props = {
      item_name: "dagger",
      item_id: 1,
      item_description: "desc",
      goldcost: 10,
      wave_lvl_drop: 10,
      rarity_name: "common",
      specials: "spec",
      abilities: "abi",
    };
    const tree = renderer
      .create(<Item {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render with props", () => {
    const props = {
      item_name: "dagger",
      item_id: 1,
      item_description: "desc",
      goldcost: 10,
      wave_lvl_drop: 10,
      rarity_name: "common",
      specials: "spec",
      abilities: "abi",
    },
      wrapper = mount(<Item {...props} />);

    expect(wrapper.props().item_name).toEqual("dagger");
    expect(wrapper.props().rarity_name).toEqual("common");
    expect(wrapper.props().goldcost).toEqual(10);
  });
});
