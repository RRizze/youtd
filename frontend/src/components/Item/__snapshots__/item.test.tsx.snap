// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item tests render correctly 1`] = `
<div
  className="item"
  data-id={1}
>
  <div
    className="item__general-info"
  >
    <div
      className="item__header"
    >
      <figure>
        <img
          alt="item icon"
          className="item__icon"
          src="#"
        />
        <figcaption>
          <h2
            className="item__name"
          >
            dagger
          </h2>
        </figcaption>
      </figure>
    </div>
    <div
      className="item__features"
    >
      <table
        className="item__table"
      >
        <thead>
          <tr>
            <th
              className="item__table-header"
              colSpan={2}
            >
              General Info
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th>
              Rarity
            </th>
            <td
              style={
                Object {
                  "color": "#ffffff",
                }
              }
            >
              common
            </td>
          </tr>
          <tr>
            <th>
              Goldcost
            </th>
            <td>
              10
            </td>
          </tr>
          <tr>
            <th>
              Wave Lvl Drop
            </th>
            <td>
              10
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="item__main-info"
  >
    <h2
      className="item__main-info-title"
    >
      Main Info
    </h2>
    <div
      className="item__description"
    >
      <h3
        className="main-info-title"
      >
        Item Description
      </h3>
      <p>
        desc
        .
      </p>
    </div>
    <div
      className="item__specials"
    >
      <h3
        className="main-info-title"
      >
        Specials: 
      </h3>
      <p>
        spec
      </p>
    </div>
    <div
      className="item__abilities"
    >
      <h3
        className="main-info-title"
      >
        Abilities: 
      </h3>
      <p>
        abi
      </p>
    </div>
  </div>
</div>
`;
